### DISCUSSION TITLE
Passing 16 bit data to AudioBuffer (as byte[])

### DISCUSSION THREAD

**Discussion by Gavin-Williams** [amerkoleci/Vortice.Windows#339] [Q&A] [answered]
AudioBuffer takes a byte[] but typically data will be 16 bit? I'm not sure how to deal with this.

Attempt 1: (language restriction)
```
SoundBuffer = new AudioBuffer(soundData); 
```
Attempt 2: (language restriction)
```
SoundBuffer = new AudioBuffer((byte[])soundData);
```
Attempt 3: (performance cost)
```
var SoundDataBytes = SoundData.Select(static value => (byte)value).ToArray();
```
Atttempt 4: (maybe, but too cumbersome)
```
            GCHandle handle = GCHandle.Alloc(soundData, GCHandleType.Pinned);
            try
            {
                IntPtr ptr = handle.AddrOfPinnedObject();
                SoundBuffer = new AudioBuffer(ptr, samples * 2);
            }
            finally
            {
                if (handle.IsAllocated)
                    handle.Free();
            }
```

What is the recommended way to pass 16 bit data in?

**Comment by amerkoleci** [comment]
Monday I'll add generic data and span support.

**Comment by amerkoleci** [comment, marked as answer]
NVM there was already present, take look here: https://github.com/amerkoleci/Vortice.Windows/blob/main/src/Vortice.XAudio2/AudioBuffer.cs#L75

**Reply by Gavin-Williams** [reply]
OK. I see, because it's unsafe, you had to create a static constructor. I was only looking at the constructors.

**Comment by Gavin-Williams** [comment]
What is wrong with this..
```cs
        // construct sound buffer
        int seconds = 5;
        int channels = 1;
        int rate = 44100;
        int samples = seconds * channels * rate;
        short[] data = new short[samples];
        for (int i = 0; i < samples; i++)
        {
            data[i] = (short)(Math.Sin(i*0.02f) * short.MaxValue);
        }
        ReadOnlySpan<short> span = new ReadOnlySpan<short>(data);
        SoundBuffer = AudioBuffer.Create(span); // ** exception thrown **
```
> Exception thrown at 0x00007FFBE185E136 (coreclr.dll) in DemoBasicApp.exe: 0xC0000005: Access violation writing location 0x0000019375003000.

**Reply by Gavin-Williams** [reply]
There is a bug in AudioBuffer.Create()
```
void* dataPtr = MemoryHelpers.AllocateMemory((nuint)data.Length);
```
should be...
```
void* dataPtr = MemoryHelpers.AllocateMemory((nuint)sizeInBytes);
```

**Reply by Gavin-Williams** [reply]
pull request: https://github.com/amerkoleci/Vortice.Windows/pull/340

**Comment by amerkoleci** [comment]
Version 2.1.26 has been released with those improvements.
